src/freenet/client/async/KeyListenerTracker.java

```diff
@@ -95,61 +102,197 @@ protected static int fixRetryCount(int retryCount) {
 		return Math.max(0, retryCount-MIN_RETRY_COUNT);
 	}
 
+	private boolean contains(KeyListener[] listeners, KeyListener listener) {
+		for(KeyListener l : listeners) {
+			if(l == listener) return true;
+		}
+		return false;
+	}
+
 	public void addPendingKeys(KeyListener listener) {
 		if(listener == null) throw new NullPointerException();
+		byte[] wantedKey = listener.getWantedKey();
+		ByteArrayWrapper wrapper = wantedKey != null ? new ByteArrayWrapper(saltKey(wantedKey)) : null;
+		assert(Arrays.equals(wantedKey, listener.getHasKeyListener().getWantedKey()));
 		synchronized (this) {
 			// We have to register before checking the disk, so it may well get registered twice.
-			if(keyListeners.contains(listener))
-				return;
-			keyListeners.add(listener);
+			if(wantedKey != null) {
+				Object o = singleKeyListeners.get(wrapper);
+				if(o == null) {
+					singleKeyListeners.put(wrapper, listener);
+				} else if(o instanceof KeyListener) {
+					if(listener == (KeyListener)o) return;
+					singleKeyListeners.put(wrapper,
+							new KeyListener[] { (KeyListener)o, listener });
+				} else {
+					@SuppressWarnings("unchecked")
+					KeyListener[] listeners = (KeyListener[])o;
+					if(contains(listeners, listener)) return;
+					KeyListener[] newListeners = Arrays.copyOf(listeners, listeners.length+1);
+					newListeners[listeners.length] = listener;
+					singleKeyListeners.put(wrapper, newListeners);
+				}
+			} else {
+				if(keyListeners.contains(listener))
+					return;
+				keyListeners.add(listener);
+			}
 		}
 		if (logMINOR)
-			Logger.minor(this, "Added pending keys to "+this+" : size now "+keyListeners.size()+" : "+listener);
+			Logger.minor(this, "Added pending keys to "+this+" : size now "+this.keyListeners.size()+"/"+singleKeyListeners.size()+" : "+listener);
 	}
 	
 	public boolean removePendingKeys(KeyListener listener) {
-		boolean ret;
+		boolean ret = false;
+		byte[] wantedKey = listener.getWantedKey();
+		ByteArrayWrapper wrapper = wantedKey != null ? new ByteArrayWrapper(saltKey(wantedKey)) : null;
 		synchronized (this) {
-			ret = keyListeners.remove(listener);
-			while(logMINOR && keyListeners.remove(listener))
-				Logger.error(this, "Still in pending keys after removal, must be in twice or more: "+listener, new Exception("error"));
+			if(wantedKey != null) {
+				Object o = singleKeyListeners.get(wrapper);
+				if(o == null) {
+					// do nothing
+				} else if(o instanceof KeyListener) {
+					if((ret = (listener == (KeyListener)o)))
+						singleKeyListeners.remove(wrapper);
+				} else {
+					@SuppressWarnings("unchecked")
+					KeyListener[] listeners = (KeyListener[])o;
+					KeyListener[] newListeners = new KeyListener[listeners.length-1];
+					int x = 0;
+					for (KeyListener l : listeners) {
+						if(listener == l) {
```

Comment:
From Eleriseth: See no point. None of KeyListener implementation overrides equals(), they
are not *supposed* to override equals(), and the use of == avoids
unneccessary null pointer checks.